### PR TITLE
[MOSIP-29962]Added sample template for developers reference usage

### PR DESCRIPTION
**Pull request by abhishek8shankar**

### CODE REVIEW

File: <mention service name>/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+## Add required path according location of file.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+## Add required path according location of file.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+## Add required path according location of file.
+ADD <mention folder name>/src/main/resources ./<mention folder name>/resources
+
+ADD configure.sh configure.sh
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Please add only one ADD label as this is just a template. Mention in comments multiple ADD can be added to the Dockerfile

File: <mention service name>/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+RUN chmod a+x configure.sh \
+## Add required path according location of jar.
+ && rm -f ./registration-client/target/registration-client-*-javadoc.jar \
+ ## Add required path according location of jar.
+ && rm -f ./registration-client/target/registration-client-*-sources.jar
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Please remove this 

Add entrypoint.sh and only the required template details

File: <mention service name>/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+# change permissions of file inside working dir
+RUN chown -R ${container_user}:${container_user} /home/${container_user}
+RUN chown -R ${container_user}:${container_user} /var/www
+
+# select container user for all tasks
+#USER ${container_user_uid}:${container_user_gid}
+
+ENTRYPOINT ["./configure.sh" ]
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Please check this to `entrypoint.sh`.

File: .github/workflows/push-trigger.yml [hunk 4]
```diff
@@ -0,0 +1,95 @@
+##  Reference for creating push-trigger.yml :
+
+---
+name: Maven Package upon a push
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Change this to `<Build-Tool>`

File: .github/workflows/push-trigger.yml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      - 1.*
+      - develop
+      - MOSIP*
+
+jobs:
+  build-maven-pre-registration:
+    uses: mosip/kattu/.github/workflows/maven-build.yml@master
+    with:
+      SERVICE_LOCATION: ./pre-registration
+      BUILD_ARTIFACT: pre-registration
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Please use `<mention service name>`

File: .github/workflows/push-trigger.yml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      OSSRH_TOKEN: ${{ secrets.OSSRH_TOKEN }}
+      GPG_SECRET: ${{ secrets.GPG_SECRET }}
+      SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK }}
+
+  publish_to_nexus:
+    if: "${{ !contains(github.ref, 'master') && github.event_name != 'pull_request' }}"
+    needs: build-maven-pre-registration
+    uses: mosip/kattu/.github/workflows/maven-publish-to-nexus.yml@master
+    with:
+      SERVICE_LOCATION: ./pre-registration
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Please use <mention service name>

File: .github/workflows/push-trigger.yml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    needs: build-maven-pre-registration
+    strategy:
+      matrix:
+        include:
+          - SERVICE_LOCATION: 'pre-registration/pre-registration-application-service'
+            SERVICE_NAME: 'pre-registration-application-service'
+            BUILD_ARTIFACT: 'pre-registration'
+          - SERVICE_LOCATION: 'pre-registration/pre-registration-batchjob'
+            SERVICE_NAME: 'pre-registration-batchjob'
+            BUILD_ARTIFACT: 'pre-registration'
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Please use <mention service name>

File: .github/workflows/push-trigger.yml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      ACTOR_DOCKER_HUB: ${{ secrets.ACTOR_DOCKER_HUB }}
+      RELEASE_DOCKER_HUB: ${{ secrets.RELEASE_DOCKER_HUB }}
+      SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK }}
+
+  sonar_analysis:
+    needs: build-maven-pre-registration
+    if: "${{  github.event_name != 'pull_request' }}"
+    uses: mosip/kattu/.github/workflows/maven-sonar-analysis.yml@master
+    with:
+      SERVICE_LOCATION: ./pre-registration
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Please use <mention service name>

File: <mention service name>/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,94 @@
+## Reference for creating an Dockerfile:
+
+---
+## mention requred version of java
+FROM openjdk:11
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: `FROM <base-image>:<base-image-tag>

File: <mention service name>/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+ENV client_repo_env=${client_repo_url}
+
+ENV artifactory_url_env=${artifactory_url}
+
+ENV host_name_env=${host_name}
+
+ENV keystore_secret_env=${keystore_secret}
+
+ENV signer_timestamp_url_env=${signer_timestamp_url}
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Add only one ENV and ARGS

File: <mention service name>/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+ARG container_user_uid=1001
+
+# can be passed during Docker build as build time environment for github branch to pickup configuration from.
+ARG container_user_gid=1001
+
+# install packages and create user
+RUN apt-get -y update \
+&& apt-get install -y unzip zip nginx \
+&& groupadd -g ${container_user_gid} ${container_user_group} \
+&& useradd -u ${container_user_uid} -g ${container_user_group} -s /bin/sh -m ${container_user}
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: RUN <Install required packages & create ${container_user/group} with group ${container_user_uid/gid>

File: <mention service name>/README.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+## <mention service name> - Change the name 
+
+Add the required src code under this directory.
```
Comment (syedsalman3753) [inline <mention service name>/README.md]: Please remove this line

Add steps to build the applications 

Add steps to build application via docker

File: <mention service name>/pom.xml [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+
+    <developers>
+        <developer>
+            <name>Mosip</name>
+            <email>mosip.emailnotifier@gmail.com</email>
+            <organization>io.mosip</organization>
+            <organizationUrl>https://github.com/mosip/<mention service name></organizationUrl>
+        </developer>
+    </developers>
+
```
Comment (syedsalman3753) [inline <mention service name>/pom.xml]: Please check with chandra regarding the url used

**Review by syedsalman3753 — CHANGES_REQUESTED** [review]
* dml and ddl directories are missing
* For documentation part, get it review by swati

File: <mention service name>/pom.xml (hunk 13 above)
Comment (ckm007) [reply <mention service name>/pom.xml]: change the Licence to MPL2

Comment (ckm007) [reply <mention service name>/pom.xml]: organizationUrl:  https://github.com/mosip   --> https://mosip.io

Comment (ckm007) [reply <mention service name>/pom.xml]: <url>https://github.com/mosip/<mention service name></url>

to

<url>https://mosip.io<mention service name></url>

File: .github/workflows/push-trigger.yml [hunk 14]
```diff
@@ -0,0 +1,92 @@
+##  Reference for creating push-trigger.yml :
+
+---
+name: <Build-Tool>
+
+on:
+  release:
+    types: [published]
+  pull_request:
+    types: [opened]
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml]: Please update pull_request

```
  pull_request:
    types: [opened, reopened, synchronize]
```

File: .github/workflows/push-trigger.yml [hunk 15]
```diff
@@ -0,0 +1,92 @@
+##  Reference for creating push-trigger.yml :
+
+---
+name: <Build-Tool>
```
Comment (syedsalman3753) [inline .github/workflows/push-trigger.yml:4]: Please add workflows for ui based modules

File: .github/workflows/tag.yml [hunk 16]
```diff
@@ -0,0 +1,35 @@
+name: Tagging of repos
```
Comment (syedsalman3753) [inline .github/workflows/tag.yml]: Please remove the tag.yaml file as we have a central workflows to tag branches

File: <mention service name>/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,58 @@
+## Reference for creating an Dockerfile:
+
+---
+## mention requred version of java
+FROM
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: `FROM <image-name>:\<tag-name\>

File: <mention service name>/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+# set working directory for the user
+WORKDIR /home/${container_user}
+
+ENV work_dir=/home/${container_user}
+
+## Add required path according location of file.multiple ADD can be added to Dockerfile.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+ADD configure.sh configure.sh
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Please remove this.

File: <mention service name>/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+# change permissions of file inside working dir
+RUN chown -R ${container_user}:${container_user} /home/${container_user}
+RUN chown -R ${container_user}:${container_user} /var/www
+
+# select container user for all tasks
+#USER ${container_user_uid}:${container_user_gid}
+
+ENTRYPOINT ["./entrypoint.sh" ]
+
+---
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: please remove --- from 56 line

File: <mention service name>/Dockerfile [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+ENV work_dir=/home/${container_user}
+
+## Add required path according location of file.multiple ADD can be added to Dockerfile.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+ADD configure.sh configure.sh
+
+# change permissions of file inside working dir
+RUN chown -R ${container_user}:${container_user} /home/${container_user}
+RUN chown -R ${container_user}:${container_user} /var/www
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: RUN chown -R ${container_user}:${container_user} \<other-directories\>

File: <mention service name>/README.md [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+The project requires <mention build tool>.
+1. Build and install:
+    ```
+    $ cd kernel
+    $ mvn install -DskipTests=true -<mention required service>.skip=true -Dgpg.skip=true
+    ```
+1. Build Docker for a service:
+    ```
+    $ cd <service folder>
+    $ docker build -f Dockerfile
```
Comment (syedsalman3753) [inline <mention service name>/README.md]: docker build . --build-arg SOURCE=mosip --build-arg COMMIT_HASH=<git-commit-hash-id> --build-arg COMMIT_ID=<git-commit-id> --build-arg BUILD_TIME=$(date +'%Y-%m-%d-%H-%M-%S') --file Dockerfile --tag "<docker-account-ID>/<service-name>:<tag-name>"

File: <mention service name>/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+# select container user for all tasks
+#USER ${container_user_uid}:${container_user_gid}
+
+ENTRYPOINT ["./entrypoint.sh" ]
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile:56]: Please add new file `entrypoint.sh` under <mention service name> directory

File: <mention service name>/Dockerfile [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+ADD configure.sh configure.sh
+
+# change permissions of file inside working dir
+RUN chown -R ${container_user}:${container_user} /home/${container_user}
+RUN chown -R ${container_user}:${container_user} /var/www
+
+# select container user for all tasks
+#USER ${container_user_uid}:${container_user_gid}
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: Please uncomment this

File: <mention service name>/Dockerfile [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+# set working directory for the user
+WORKDIR /home/${container_user}
+
+ENV work_dir=/home/${container_user}
+
+## Add required path according location of file.multiple ADD can be added to Dockerfile.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
+
+ADD configure.sh configure.sh
+
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile:48]: ADD entrypoint.sh $work_dir/entrypoint.sh

RUN chmod +x $work_dir/entrypoint.sh

File: <mention service name>/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+# install packages and create user
+RUN <Install required packages & create ${container_user/group} with group ${container_user_uid/gid> \
+
+# set working directory for the user
+WORKDIR /home/${container_user}
+
+ENV work_dir=/home/${container_user}
+
+## Add required path according location of file.multiple ADD can be added to Dockerfile.
+ADD <mention service name>/<mention folder name>/target ./<mention folder name>/target
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: ADD \<mention service name\>/\<mention folder name\>\/target .\/\<mention folder name\>/

File: <mention service name>/Dockerfile [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+ARG BUILD_TIME
+
+LABEL source=${SOURCE}
+LABEL commit_hash=${COMMIT_HASH}
+LABEL commit_id=${COMMIT_ID}
+LABEL build_time=${BUILD_TIME}
+
+ARG
+
+ENV
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: \# You can add multiple the environment (ENV) and argument variables (ARG).

ARG \<arg_name\>=arg-name

ENV \<env-variable-name\>="env-var-value"

**Review by syedsalman3753 — CHANGES_REQUESTED** [review]

File: <mention service name>/Dockerfile [hunk 27]
```diff
@@ -1,8 +1,7 @@
 ## Reference for creating an Dockerfile:
 
----
 ## mention requred version of java
-FROM
+FROM <tag-name>
```
Comment (syedsalman3753) [inline <mention service name>/Dockerfile]: FROM \<docker-image-name\>:\<tag-name\>

**Review by syedsalman3753 — CHANGES_REQUESTED** [review]

**Review by syedsalman3753 — APPROVED** [review]

**Pull request merged**